### ISSUE TITLE
Request timeout: `addresses/[address]/transactions?page=1`

### ISSUE THREAD

**Issue by simerplaha** [alephium/explorer-backend#300]
This page is slow to respond. 

[addresses/[address]/transactions?page=1](https://mainnet-backend.alephium.org/addresses/134dmgh9AZNimp9ZAYpWBTwv3ddouETBTyq6134mBvVk2/transactions?page=1)

Timeout in 20 seconds

![image](https://user-images.githubusercontent.com/1773953/180113716-80e12367-9a10-4020-a9c8-974b49a75e19.png)

**Comment by LeeAlephium** [comment]
Similar issue for getting the transaction count https://explorer.alephium.org/#/addresses/14kSoi8pFMn2wMJFZJgGpFSA7GyNMtwWKnsKc1KidZ8Bu

**Comment by simerplaha** [comment]
That's useful. Thank you 👍🏼

**Comment by tdroxler** [comment]
The latest `v1.8.0-rc1` with lots of optimization is available to test here: https://mainnet-explorer-v18-0-backend.alephium.org but we can't have the front-end yet, as it has many breaking changes in the api.
FE issues is here https://github.com/alephium/explorer/issues/111

**Comment by simerplaha** [comment]
Benchmarking locally shows the performance by Lee's provided link to be kinda expected. 

See the two mid benchmarks `b_noJoin` and `c_flipped` (ignore the others)
- `b_noJoin` - Current query - `WHERE main_chain = true AND address = $address`
-  `c_flipped` - Flips the order of `and` - `WHERE address = $address AND main_chain = true`

## Bench - 10,000 transactions per address

- 500 unique addresses
- 10,000 transactions per address.
- 5,000,000 total rows in `transaction_per_addresses`
![Mode Cnt Score Error Units](https://user-images.githubusercontent.com/1773953/182524187-17d34969-f4c3-4127-84a8-56679951caa6.png)

## Bench - 100,000 transactions per address

- 50 unique addresses
- 100,000 transactions per address.
- 5,000,000 total rows in `transaction_per_addresses`
![Mode Cnt Score Error Units](https://user-images.githubusercontent.com/1773953/182524318-a55a77d4-13de-4b6e-baec-b3c40ab89063.png)

## Bench - 700,000 transactions per address (similar to Lee's provided link)

- 7 unique addresses
- 700,000 transactions per address.
- 4,900,000 total rows in `transaction_per_addresses`

![Mode Cnt Score Error Units](https://user-images.githubusercontent.com/1773953/182524349-b4eef583-9e44-4dd1-9b70-7e9d288dc32d.png)

## Notes

The performance worsens with more addresses that have large number of transactions for the same number of rows in each benchmark. 

These benchmarks are on my local machine. In production, Postgres being on a remote machine, performance would plummet noticeably further (20 second timeout makes sense) .. Doesn't look good for the future.

Simple solution: Maintain another table that stores transaction count for each address during sync? Being a simple table it would be easy to cache `CaffeineCache[Address, Int] //address -> count`?  (Think this is something you'd prefer?)

These queries are just running `count`. I'd assume fetching the actual rows would have the same story.

If you are curious, the `id` benchmarks are for when the long address `String` all have an associated `id: Integer`. I've done these benchmarks before [here](https://github.com/alephium/explorer-backend/issues/130) but they might not be that impressive to sell.

**Comment by simerplaha** [comment]
Given this a fair amount of thought. I'd like to bring attention to the root problem here which is - _queries run slow with more data in a single table_. Postgres creates one file per table. We all know concurrent IO to a single file is slow. So a single large transaction table is only going to get slower and slower over time with more and more transactions. 

I've even run the above benchmarks with my own SQL library with a much simple table and single index, the outcome is the same. 

There are some obvious tricks like sharding, partitioning to improve queries but these should be last resort. 

Our current approach to caching which is "selective-caching" is not going to bring any dramatic performance improvements either. By selective-caching I mean, we pick what we want cached like latest blocks & transactions etc. and are ok if output is not always 100% accurate. We also drop the cache based on some timeout and not based on some logical reason (cost = CPU time). Following this approach, overtime, we will selectively cache more and more things which improve performance at some places and not at others. The underlying problem of overall DB performance will still exist. 

There are some other obvious things we can do to improve overall DB performance
- Split data at DB level - hot data and cold data
- Bring Postgres as close as possible to the application to cut latency cost
- Manually split data. For eg: in the above case create new tables dynamically for users with larger number of transactions. Gonna be a bit of a maintenance burden. 
- Create views/duplicate data - Create views that map directly to some UI, like a paginated table. Queries a simpler therefore faster and easy to cache using Caffeine. 
- Go the Redis route a slap an expensive cache in front.
- Experiment with Postgres level settings like `VACCUM`, buffers etc. This needs a bit of experimenting.
- Remove things that are considered as anti-pattern i.e. polling and concurrent batch IO
- Possibly others …

Just a personal opinion: Being a storage-engine developer I'm a bit biased towards LSM trees and B-trees truly suck. Prefer embedded storage-engines over years old solutions like Postgres that don't scale well on single machine and are not suitable for modern data requirements and are expensive in every way. I'm not going to try to convince anyone to move away from Postgres either. We've had this conversation once before, people have strong opinion on this stuff. Let's move on.

In summary, please think about the underlying problem here and think how we prefer to solve it. Because what we are doing right now to improve performance like selective-caching, micro-optimisations is not going to do any miracles. 

The number of end solutions are unlimited. Anything can be implemented right, it's just code! So the actual problem here is what's our preferred approach? We’ve been through this before and to save time I trying not to implement something that is not preferred.

Even if our preferred solution is sticking with selective-caching at least we should know that we are not expecting sudden dramatic performance boost. 

Thoughts?

**Comment by tdroxler** [comment]
Back in time when we discussed about Postgres, it was clear that the bottleneck was not (only) Postgres itself, but mostly our queries and indexes. With all our latest optimizations, we now see the limitation of PG and it's time we re-discuss the subject. 

When we started the explorer-backend, it was clear (at least for me :stuck_out_tongue_closed_eyes: ) that we needed a SQL database, as the all models were really suited for that (joining input/outputs etc) and we also got inspired by [Ergo](https://explorer.ergoplatform.com/en/) which was our reference and they do use postgres, but their explorer is now also slow and we already have much more blocks than they do.

A lot of our latest optimizations were to remove all our joins to have faster queries, which is the case, but still PG is slow....

We would need to investigate, but it might be possible that we could even get rid of a SQL-type database.

I really don't have much experiences with DBs so I don't have strong opinion of the best one to choose, but I'm wondering if there is such DB that won't suffer from a table with millions of rows? Thinking out loud, but what's costly with our current PG is that we always need to `ORDER BY timestamp` to have the latest 20 transactions of a user out of its millions txs, maybe a DB that is already ordered could help? because that's a bit the case for every explorer-FE page, we don't have that many information on one page, but still the query has to deal with EVERY data of an address. Same for the list of blocks in the home page, we need to sort them (arg).

Anyway I wouldn't rush on a new DB as the current state is fine for "normal" address and our new AWS instance is doing much better without all the disk swap, even the addresses on top of this discussion are doing good. Maybe we should take some time to lay down what are the top requirement for our data and queries to be fast? Maybe you already have some ideas @simerplaha ? I'm also bringing @killerwhile in the discussion as he might have some ideas too.

Going back on the list you proposed, some ideas are good, but in the end no matter what we do, PG is and will stay slow for some of our needs, even like a stupid `select count` (you already know that, but I'm slow as PG to understand stuff ^^)

**Comment by simerplaha** [comment]
> When we started the explorer-backend, it was clear (at least for me 😝 ) that we needed a SQL database ...

Totally, SQL seems to be the first choice for all. Seems like a safe bet.

> but I'm wondering if there is such DB that won't suffer from a table with millions of rows?

Maybe MyRocks. It's MySQL with RocksDB. Being SQL, it should work with our code with none or minimal change. I haven't followed this project's progress in over a year so I'm not too sure. Still gotta do the due diligence - benchmarks etc ...

> Anyway I wouldn't rush on a new DB as the current state is fine for "normal" address ...

Sounds good. 👍🏼

> Maybe we should take some time to lay down what are the top requirement for our data and queries to be fast?

Yep totally!

> Maybe you already have some ideas @simerplaha ? 

Yep. Can get back to this whenever you think the time is right. Like you said - "I wouldn't rush on a new DB ... "

**Comment by tdroxler** [comment]
The new AWS instance improved a lot the performance, we can close this for now and we'll create a task about the investigation of a new DB
